core/src/main/java/org/apache/iceberg/SnapshotManager.java

```diff
@@ -0,0 +1,152 @@
+        "Cannot cherry pick unknown snapshot id: %s", snapshotId);
+
+    operation = SnapshotManagerOperation.CHERRYPICK;
+    this.targetSnapshotId = snapshotId;
+
+    // Pick modifications from the snapshot
+    Snapshot cherryPickSnapshot = base.snapshot(this.targetSnapshotId);
+    String wapId = stagedWapId(cherryPickSnapshot);
+    boolean isWapWorkflow =  wapId != null && !"".equals(wapId);
+    if (isWapWorkflow) {
+      if (base.isWapIdPublished(wapId)) {
+        throw new DuplicateWAPCommitException(wapId);
+      }
+    }
+    // only append operations are currently supported
+    if (!cherryPickSnapshot.operation().equals(DataOperations.APPEND)) {
+      throw new UnsupportedOperationException("Can cherry pick only append operations");
+    }
+    // this is to ensure we add files only once for each targetSnapshotId, to protect from
+    // duplicate additions if commit retries the same cherrypick operation on failure.
+    if (!snapshotsAlreadyCherrypicked.contains(targetSnapshotId)) {
+      for (DataFile addedFile : cherryPickSnapshot.addedFiles()) {
+        add(addedFile);
+      }
+      // this property is set on target snapshot that will get published
+      if (isWapWorkflow) {
+        set(SnapshotSummary.PUBLISHED_WAP_ID_PROP, wapId);
+      }
+      snapshotsAlreadyCherrypicked.add(targetSnapshotId);
+    }
+
+    return this;
+  }
+
+  @Override
+  public ManageSnapshots setCurrentSnapshot(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot roll back to unknown snapshot id: %s", snapshotId);
+
+    operation = SnapshotManagerOperation.ROLLBACK;
+    this.targetSnapshotId = snapshotId;
+
+    return this;
+  }
+
+  @Override
+  public ManageSnapshots rollbackToTime(long timestampMillis) {
+    operation = SnapshotManagerOperation.ROLLBACK;
+    // find the latest snapshot by timestamp older than timestampMillis
+    Snapshot snapshot = SnapshotUtil.findLatestSnapshotOlderThan(base, timestampMillis);
+    Preconditions.checkArgument(snapshot != null,
+        "Cannot roll back, no valid snapshot older than: %s", timestampMillis);
+    this.targetSnapshotId = snapshot.snapshotId();
+
+    return this;
+  }
+
+  @Override
+  public ManageSnapshots rollbackTo(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot roll back to unknown snapshot id: %s", snapshotId);
+
+    ValidationException.check(SnapshotUtil.isCurrentAncestor(base, snapshotId),
+        "Not a valid snapshot to rollback to. Cannot rollback to a snapshot " +
+            "that's not an ancestor of the current snapshot.");
+    return setCurrentSnapshot(snapshotId);
+  }
+
+  @Override
+  public Snapshot apply() {
+    // common checks
+    ValidationException.check(targetSnapshotId != null,
+        "Cannot run operations on an unknown version: call setCurrentSnapshot, rollbackAtTime or cherrypick");
```

Comment:
If we detect that there are no changes specified, then why not just return `currentSnapshot`? Most other operations work that way, where if you don't configure them then committing is a noop instead of failure.

Reply:
done.